packages/types/src/codec/Base.ts

```diff
@@ -42,7 +43,11 @@ export abstract class Base<T extends Codec> implements Codec {
    * @description returns a hash of the contents
    */
   public get hash (): CodecHash {
-    return this.registry.hash(this.toU8a());
+    if (!this.#hashCache) {
```

Comment:
This one is actually quite problematic, specifically for the case you want to optimize, e.g. extrinsics. The issue is the following - especially in the case of extrinsics, the values are mutable. Basically you can create a extrinsic via `api.tx....` which will have one hash and then after submission (signature attached), it will have a different hash.

So where this will fail -

```js
const tx = api.tx.balances.transfer(BOB, 12345);

console.log('pre signing hash', tx.hash.toHex());

await tx.signAndSend(...);

// in this case the object mutated, the hash should be is different, be here showed as the same
console.log('post signing hash', tx.hash.toHex());
```

Reply:
The way around this is probably to actually check the extrinsic differently - i.e. rather use the `.eq` on the extrinsic to see if all the details do match. (Once again, as soon as something mismatched is found, it will bail immediately)

This is probably a better approach since we won't serialize and hash, which is expensive.

Reply:
Hm, but we don't always have access to the full extrinsic to check, this is why extrinsics are checked by hash instead.

What if hash is cached in an extrinsic instead and cache is reset on signature, will that work?

Reply:
Yes, that works - it makes sense to move this logic to the extrinsic itself (since that is where you are trying to optimize).